subprojects/enterprise-operations/src/main/java/org/gradle/execution/work/WorkExecutionBuildOperationType.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.work;
+
+import org.gradle.internal.operations.BuildOperationType;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Captures the execution of a unit of work. This includes all operations related to Gradle's bookkeeping and optimizations.
+ * Among other things it includes snapshotting inputs and outputs, up-to-date checks, build cache operations.
+ * It also wraps the {@link WorkActionExecutionBuildOperationType execution of the action}.
+ *
+ * The operations can be nested, i.e. when a task executes a worker API action.
+ *
+ * @see WorkActionExecutionBuildOperationType
+ */
+public interface WorkExecutionBuildOperationType extends BuildOperationType<WorkExecutionBuildOperationType.Details, WorkExecutionBuildOperationType.Result> {
+
+    interface Details {
+
+        /**
+         * The type of work represented by the build operation.
+         */
+        Class<?> getWorkType();
+
+        /**
+         * Path to the owning build.
+         */
+        String getBuildPath();
+
+        /**
+         * A textual identifier for the unit of work when available.
+         *
+         * Tasks will expose their task paths here. Artifact transforms and workers don't expose anything here.
+         */
+        Optional<String> getIdentity();
+
+        /**
+         * A unique number identifying the unit of work being executed in the context of the current build.
+         */
+        long getUniqueId();
+
+        /**
+         * The type implementing the unit of work.
+         */
+        Class<?> getImplementation();
+    }
+
+    interface Result {
+        /**
+         * Whether the unit of work had any actions. Lifecycle tasks don't have actions.
+         *
+         * @see org.gradle.api.internal.tasks.TaskStateInternal#isActionable()
+         */
+        boolean isActionable();
```

Comment:
Why are there all these task specific methods in the `Result`? Shouldn't the result be an empty interface?

Reply:
There are two possible approaches here:

1) capture everything that might be relevant to any work type in the new operations,
2) capture only the things applying to everything in the new types, and let additional build operations specific to each work type capture the specific data.

I chose to go with 1) in the first draft of this proposal, essentially copying stuff from `ExecuteTaskBuildOperationType`. 

The other option would be to capture only data related to caching and incrementality here, and let task path and whether or not the thing is actionable be captured in a task-specific operation (perhaps we can keep `ExecuteTaskBuildOperationType` for this purpose).

> Shouldn't the result be an empty interface?

Do you propose to move the caching/incrementality stuff to separate build operations as well, even if they apply to all kinds of units of work? While that seems more _correct,_ I'm somewhat worried that adding complexity can lead to us not finishing this.

@ldaley which approach do you think is better here?